### ISSUE TITLE
It's unclear how/where I should configure job concurrency in dagster OS based on the documentation

### ISSUE THREAD

**Issue by JasperHG90** [dagster-io/dagster#19402]
### What's the issue or suggestion?

*Situation*
I have deployed dagster OS to better understand the deployment process. I have separate deployments for the dagster daemon & server and my code locations.

I am deploying a dag with a single partitioned asset.

*What I want to do*
I want to control the concurrency of the job to which the asset belongs, ideally on a code-location basis.

*Documentation I'm following*
I'm following the article ["Limiting concurrency in data pipelines"](https://docs.dagster.io/guides/limiting-concurrency-in-data-pipelines). I'm having a hard time understanding exactly **where** I need to specify limits on the concurrency of my job.

For example, the [following section](https://docs.dagster.io/guides/limiting-concurrency-in-data-pipelines#limiting-overall-runs):

<img width="1034" alt="image" src="https://github.com/dagster-io/dagster/assets/8079803/ee757dfe-4fdc-453b-be61-818a6088757c">

This makes it seem as if you can control the concurrency on a code-location basis using the `dagster.yaml` file. While this works when setting the `DAGSTER_HOME` environment variable and executing `dagster dev`, you cannot use this when deploying your code location (which calls `dagster api grpc`).  See also [this issue](https://github.com/dagster-io/dagster/discussions/19119).

After some searching, I found that the helm `values.yaml` file [contains the same section](https://github.com/dagster-io/dagster/blob/master/helm/dagster/values.yaml#L1067-L1084) as mentioned in the docs. This suggests that, when using dagster OS, you need to place the concurrency limits here.

It would be great if the docs were clearer on this.

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

**Comment by sam-goodwin** [comment]
I am following the docs precisely and getting errors. It's very confusing to understand.

```py
my_job = define_asset_job(
    name="my_jb",
    partitions_def=scan_id,
    executor_def=multiprocess_executor,
    config={
        "execution": {
            "config": {
                "multiprocess": {
                    "tag_concurrency_limits": [
                        {
                            "key": "yarn",
                            "value": "dask",
                            "limit": 3,
                        }
                    ],
                },
            }
        }
    },
)
```

Error:
```
Backfill failed for jnpeozgr: dagster._core.errors.DagsterInvalidConfigError: Error in config for job orion_human
    Error 1: Received unexpected config entry "multiprocess" at path root:execution:config. Expected: "{ max_concurrent?: Int? retries?: { disabled?: { } enabled?: { } } start_method?: { forkserver?: { preload_modules?: [String] } spawn?: { } } tag_concurrency_limits?: [{ key: String limit: Int value?: (String | { applyLimitPerUniqueValue: Bool }) }] }".

```

Docs:
![image](https://github.com/dagster-io/dagster/assets/38672686/d3308ab8-ff28-4dde-a5bf-560cde0bfe16)

**Comment by garethbrickman** [comment]
@sam-goodwin Could you try restructuring your code as below? 

```python
# Configure the multiprocess_executor with the desired settings
configured_multiprocess_executor = multiprocess_executor.configured(
    {
        "max_concurrent": 3,  # limits concurrent processes to 3
        "tag_concurrency_limits": [
            {
                "key": "yarn",
                "value": "dask",
                "limit": 3,
            }
        ],
    }
)

# Define the job with the configured executor
my_job = define_asset_job(
    name="my_job",
    partitions_def=scan_id,
    executor_def=configured_multiprocess_executor,
)
```

**Comment by axellpadilla** [comment]
This is the same for the dagster.yaml, very confuse, we need a complete example

**Comment by garethbrickman** [comment]
We have a new guide for managing concurrency in our forthcoming new docs. Can preview its content in this PR: https://github.com/dagster-io/dagster/pull/24296

**Comment by garethbrickman** [comment]
We have a new guide for managing concurrency in our forthcoming new docs. Can preview it here: https://docs-preview.dagster.io/guides/managing-concurrency

**Comment by axellpadilla** [comment]
> We have a new guide for managing concurrency in our forthcoming new docs. Can preview it here: https://docs-preview.dagster.io/guides/managing-concurrency

Some new useful info there, but reading trough it I realize that I would still have no clue of the `dagster.yaml` structure if it were my first time there. I needed to check the code, for example, on my config I added the code dirs so I can read them if needed, those modules where a lot more helpful structuring my dagster.yaml but I'm still not sure it is 100% correct, an official updated and complete commented file should be available:

```yaml
# dagster\_core\instance\config.py
storage:
  postgres:
    postgres_db:
      username: { env: DAGSTER_PG_USERNAME }
      password: { env: DAGSTER_PG_PASSWORD }
      hostname: { env: DAGSTER_PG_HOST }
      db_name: { env: DAGSTER_PG_DB }
      port: { env: DAGSTER_PG_PORT }
auto_materialize:
  use_sensors: true
# Opt in to run monitoring
run_monitoring:
  enabled: true
  # values below are the defaults, and don't need to be specified except to override them
  start_timeout_seconds: 180
  cancel_timeout_seconds: 180
  max_resume_run_attempts: 0 # experimental if above 0
  poll_interval_seconds: 120
  free_slots_after_run_end_seconds: 300 # free any hanging concurrency slots after 5 minutes from the end of a run

# run_coordinator: # for custom coordinators
#   module: dagster.core.run_coordinator
#   class: QueuedRunCoordinator
#   config:
run_queue: #dagster\_core\run_coordinator\queued_run_coordinator.py
  max_concurrent_runs: 8 #  10 default, -1 unlimited, 0 no runs
  tag_concurrency_limits:
    # - key: "database"
    #   value: "dwh" # applies when the `database` tag has a value of `redshift`
    #   limit: 8
    - key: "dagster/backfill" # applies when the `dagster/backfill` tag is present, regardless of value
      limit: 8
    # - key: "use-case"
    #   value:
    #     applyLimitPerUniqueValue: true
    #   limit: 3
  block_op_concurrency_limited_runs:
    enabled: true
    op_concurrency_slot_buffer: 1 # buffer of 1 means 1 run might be started even though there are no slots available
  dequeue_use_threads: true
  dequeue_num_workers: 8

concurrency:
  default_op_concurrency_limit: 1
```

Here browsing the code for example, are some hidden configs, like the use_threads an num_workers on auto_materialize, I couldn't even find on docs or examples online.

**Comment by albertkim** [comment]
I'd like to 👍 this. I'm trying to set up global concurrency limits for some of my assets, and it's extremely unclear where in the dagster.yml file I need to add my concurrency keys. A single complete example like what axellpadilla posted above is necessary.

**Comment by shanirosen-airis** [comment]
> We have a new guide for managing concurrency in our forthcoming new docs. Can preview it here: https://docs-preview.dagster.io/guides/managing-concurrency

Getting 404 on this :( Facing similar issues, the documentation is unclear on this

**Comment by garethbrickman** [comment]
@shanirosen-airis The guide was moved within the site, new URL: https://docs-preview.dagster.io/guides/operate/managing-concurrency

**Comment by lcendoya** [comment]
I have set up concurrency limits with this code:
```
@asset(
    partitions_def=client_partitions,
    pool="activities_writes",
    deps=["dlt_rest_api_activities"],
    group_name="sentinel_one",
    automation_condition=AutomationCondition.eager()
)
def activities(context: AssetExecutionContext):
```
and also set up a `dagster.yml` with this:
```
concurrency:
  pools:
    granularity: 'run'
    default_limit: `1`
```
However, when I tried to materialize runs, they all remained in queued status. I could only fix this by manually adding the pool limit through the UI in deployment --> concurrency, the queue problem was fixed. So apparently setting `default_limit: 1` through `dagster.yml` alone doesn't work

**Comment by themightychris** [comment]
The docs are all over the place on this now.

The [`dagster.yaml` docs](https://docs.dagster.io/deployment/oss/dagster-yaml#full-configuration-specification) have both of these in the example:

<img width="305" height="182" alt="Image" src="https://github.com/user-attachments/assets/00faf2d7-b3e0-4568-8893-283bf19505d4" />

<img width="382" height="225" alt="Image" src="https://github.com/user-attachments/assets/569388d8-8075-4a1f-94ff-63d6eb45117c" />

And then the [managing concurrency guide](https://docs.dagster.io/guides/operate/managing-concurrency) has this:

<img width="969" height="258" alt="Image" src="https://github.com/user-attachments/assets/a02f5153-0f46-4d93-972a-35f8d75adb80" />

----

Are they redundant? Is one deprecated? Do they apply in different conditions?

**Comment by ArcticXWolf** [comment]
I agree with @themightychris , the docs are conflicting on this. Maybe the Managing Concurrency Guide should be updated with the following structure:

* A general paragraph explaining which options for managing concurrency you have and on what levels (global, code location, job, step, etc).
* Which concurrency options modify behavior for which levels (with configuration examples or the full reference)
* Add the missing options that @themightychris mentioned (runqueue.max_concurrent_run and run_coordinator.config..)

Sadly I dont know the difference either (and would love to hear an answer to this).
Who could know more? @garethbrickman  or @kingsley-123  maybe? (just mentioning you two here since you made the docs changes last time and might know more about this)

**Comment by axellpadilla** [comment]
I crafted this using Copilot making it check source code and re-check an iterate verification—does this look precise? It seems like each code location developer needs to collaborate with the instance‑level owner to achieve full concurrency control.

---

## Summary: Concurrency levels and configuration surfaces

Concurrency in Dagster is controlled at **four conceptual levels**, exposed through **three main configuration surfaces**:

1. **Instance / deployment level** (global to the Dagster instance)
2. **Code‑location / job level** (per run)
3. **Per‑op / per‑asset level**
4. **UI / GraphQL surface** (introspection and editing of instance‑level settings)

In Dagster **OSS**, the instance‑level configuration lives in the instance’s `dagster.yaml` (or is generated from Helm `values.yaml` for the daemon/webserver). **Code locations** (pods/containers running `dagster api grpc`) do *not* own instance‑wide concurrency; they only define jobs/ops and per‑run executor config. So yes: **code‑location developers and the instance operator must coordinate** on concurrency settings.

Below is how each level maps to concrete config options.

---

## 1. Instance / deployment level (global to the Dagster instance)

These settings apply **across all code locations** attached to an instance. They are read by the instance process (daemon / webserver), not by code locations.

### 1.1 Global run concurrency

Controls how many runs are allowed to be “in progress” at once, optionally per tag. These all ultimately feed into the **`QueuedRunCoordinator` / run queue**.

**New style: `concurrency.runs.*`**

Configured in instance config:

```yaml
concurrency:
  runs:
    max_concurrent_runs: 15
    tag_concurrency_limits:
      - key: 'database'
        value: 'redshift'
        limit: 5
```

- `concurrency.runs.max_concurrent_runs`  
  Max number of runs in progress at once.
- `concurrency.runs.tag_concurrency_limits`  
  Limits based on run tags, e.g. only N runs with a given tag or per unique tag value.

These are merged into the run queue configuration in code:

```python
RunQueueConfig(
  max_concurrent_runs=run_settings.get("max_concurrent_runs", self.max_concurrent_runs),
  tag_concurrency_limits=run_settings.get("tag_concurrency_limits", self.tag_concurrency_limits),
  ...
)
```

**Legacy: `run_queue.*`**

Older style documented in `dagster.yaml`:

```yaml
run_queue:
  max_concurrent_runs: 15
  tag_concurrency_limits:
    - key: 'database'
      value: 'redshift'
      limit: 4
    - key: 'dagster/backfill'
      limit: 10
```

- `run_queue.max_concurrent_runs`
- `run_queue.tag_concurrency_limits`

These configure the same underlying run queue behavior, but **conflict** with `concurrency.runs.*`. If `concurrency.runs` is present, Dagster raises a config error when these `run_queue` fields are set.

**Direct run coordinator config: `run_coordinator.config.*`**

Directly configures the `QueuedRunCoordinator`:

```yaml
run_coordinator:
  module: dagster.core.run_coordinator
  class: QueuedRunCoordinator
  config:
    max_concurrent_runs: 25
    tag_concurrency_limits:
      - key: 'database'
        value: 'redshift'
        limit: 5
```

- `run_coordinator.config.max_concurrent_runs`
- `run_coordinator.config.tag_concurrency_limits`

In Helm, this is set via:

```yaml
dagsterDaemon:
  runCoordinator:
    type: QueuedRunCoordinator
    config:
      queuedRunCoordinator:
        maxConcurrentRuns: 25
        tagConcurrencyLimits:
          - key: database
            value: redshift
            limit: 5
```

which is rendered into `run_coordinator.config` in the generated `dagster.yaml`.

**Takeaway:** All of the above are different *surfaces* for configuring **the same logical “run queue / run coordinator” limits**. Modern docs prefer `concurrency.runs.*`, but `run_queue.*` and `run_coordinator.config.*` still exist and must be reconciled.

---

### 1.2 Concurrency pools (op/asset limits across runs)

Concurrency pools control **how many ops/assets can be executing concurrently across all runs**, optionally at “run” vs “op” granularity.

Configured at instance/deployment level:

```yaml
concurrency:
  pools:
    default_limit: 1
    granularity: "run"             # or "op"
    op_granularity_run_buffer: 10
```

Key options:

- `concurrency.pools.default_limit`  
  Default max concurrent operations for an unconfigured pool.
- `concurrency.pools.granularity` = `op` or `run`  
  - `op`: limit is “number of ops in progress”.
  - `run`: limit is “number of runs in progress that include ops in this pool”.
- `concurrency.pools.op_granularity_run_buffer`  
  Buffer for how many runs can be launched even if all their steps would be blocked by pool limits (when `granularity=op`).

Pools themselves are then *referenced* by ops/assets in code (see Per‑op / per‑asset level).

---

## 2. Code‑location / job level (per run)

These settings live with job definitions and are applied to **a single run** at a time, not globally.

### 2.1 Executor concurrency (`max_concurrent`)

Controls how many ops are allowed to execute concurrently **inside one run**, based on the executor.

Example: Kubernetes job executor:

```python
# in dagster_k8s.executor.py

`max_concurrent` limits the number of pods that will execute concurrently for one run. By default
there is no limit – it will maximally parallel as allowed by the DAG. Note that this is not a
global limit.
```

You set this in job/run config:

```yaml
execution:
  config:
    max_concurrent: 4
```

or via Python config for the executor. This is:

- Per run (job‑level).
- Owned by the **code location**.
- Independent of global run limits (`concurrency.runs.*`) and pools.

---

## 3. Per‑op / per‑asset level

On this level, you **assign resources to constraints** rather than defining the constraints themselves.

Two main tools:

- **Pools**: Assign ops or assets to named concurrency pools in code so that they obey `concurrency.pools.*` limits at the instance level.
- **Tags**: Add run tags (e.g. `database=redshift`, `dagster/backfill=…`) to jobs/runs so that they obey `concurrency.runs.tag_concurrency_limits`.

These are defined in the code location (e.g. decorators, tags, metadata), but they take effect because of **instance‑level** limits.

---

## 4. UI / GraphQL surface

The Dagster UI and GraphQL API let you **view and sometimes edit** the instance‑level concurrency configuration.

A key type is `RunQueueConfig` in GraphQL:

```ts
type RunQueueConfig = {
  maxConcurrentRuns: number;
  tagConcurrencyLimitsYaml: string | null;
};
```

- The “Instance > Concurrency” UI displays:
  - `Max concurrent runs`
  - `Tag concurrency limits`
- These values ultimately come from the instance config (`concurrency.runs`, `run_queue`, or `run_coordinator.config`, depending on how you configured it).

So the UI is **not yet another concurrency layer**; it is a way to introspect and adjust the instance‑level configuration that the daemon is enforcing.

---

## Collaboration implication

Given this structure, your conclusion is accurate:

- **Code‑location developers**:
  - Control per‑run executor settings (`max_concurrent`).
  - Assign ops/assets to pools and run tags in code.
- **Instance / platform owners**:
  - Control global run limits (`concurrency.runs.*`, legacy `run_queue.*`, `run_coordinator.config.*`).
  - Control concurrency pools (`concurrency.pools.*`) and their limits.

To achieve a coherent concurrency strategy, **they must collaborate**:

- Developers decide *which* jobs/ops/assets should be constrained and how (pools, tags, per‑run executor limits).
- Instance managers decide *how strict* those limits are at the deployment level.

This matches the actual behavior in the source (run coordinator, instance config, Helm templates, and docs).

**Comment by ArcticXWolf** [comment]
@axellpadilla Thanks, but I still think the docs should be cleared up on this
